android/mobile/app/src/main/java/com/ttd/sso/sample/ui/AuthScreenViewModel.kt

```diff
@@ -0,0 +1,127 @@
+import android.content.Intent
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import com.myopenpass.auth.OpenPassManager
+import com.myopenpass.auth.OpenPassManagerState
+import com.myopenpass.auth.flow.WebSignInFlowClient
+import com.myopenpass.auth.flow.WebSignInFlowState
+import com.ttd.sso.sample.ui.AuthScreenAction.CheckSignIn
+import com.ttd.sso.sample.ui.AuthScreenAction.SignIn
+import com.ttd.sso.sample.ui.AuthScreenAction.SignOut
+import com.ttd.sso.sample.ui.AuthScreenState.ErrorState
+import com.ttd.sso.sample.ui.AuthScreenState.SignedInState
+import com.ttd.sso.sample.ui.AuthScreenState.SignedOutState
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+sealed interface AuthScreenAction : ViewModelAction {
+    data class SignIn(val context: Context) : AuthScreenAction
+    data class CheckSignIn(val intent: Intent) : AuthScreenAction
+    data object SignOut : AuthScreenAction
+}
+
+sealed interface AuthScreenState : ViewState {
+    data class SignedInState(val email: String) : AuthScreenState
+    data object SignedOutState : AuthScreenState
+    data class ErrorState(val error: Throwable) : AuthScreenState
+}
+
+class AuthScreenViewModel : BasicViewModel<AuthScreenAction, AuthScreenState>() {
+
+    private val _viewState = MutableStateFlow<AuthScreenState>(SignedOutState)
+    override val viewState: StateFlow<AuthScreenState> = _viewState.asStateFlow()
+
+    private var client: WebSignInFlowClient? = null
+    private val manager =  OpenPassManager.getInstance()
+
+    init {
+        // Observe the state of the OpenPassManager and translate those into our own ViewState.
+        viewModelScope.launch {
+            manager.state.collect { state ->
+                Log.d(TAG, "State Update: $state")
+
+                when (state) {
+                    is OpenPassManagerState.Loading -> Unit
+                    is OpenPassManagerState.SignedOut -> _viewState.emit(SignedOutState)
+                    is OpenPassManagerState.SignedIn -> state.tokens.idToken?.email?.let { _viewState.emit(SignedInState(it)) }
+                    is OpenPassManagerState.Error -> {
+                        _viewState.emit(ErrorState(state.error))
+                    }
+                }
+            }
+        }
+    }
+
+    override fun processAction(action: AuthScreenAction) {
+        Log.d(TAG, "Action: $action")
+
+        viewModelScope.launch {
+            when (action) {
+                is SignIn -> signIn(action.context)
+                is CheckSignIn -> checkSignIn(action.intent)
+                is SignOut -> signOut()
+            }
+        }
+    }
+
+    private suspend fun signIn(context: Context) {
+        client = WebSignInFlowClient(CLIENT_ID, manager)
+        client?.let { client ->
```

Comment:
```suggestion
        client = WebSignInFlowClient(CLIENT_ID, manager).also { client ->
```

Reply:
But `client` here is the param in the also block, it won't be a problem, it won't be null or unassigned.

To avoid shadowing, you might call it newClient or whatever and swap all usages in the also for that. It doesn't seem to be calling other methods that assume the `public val client` is in there, at least I can't see them easily.

Reply:
>But client here is the param in the also block, it won't be a problem, it won't be null or unassigned.

But the client variable is used *outside* this block, in `checkSignIn`. I have debugged it after making this change, and basically the variable never gets set, so when an intent comes into the app, and `checkSignIn` is executed, there is no client to try and handle it.

Reply:
so if you want to only put it there after initializing, how about 
```
WebSignInFlowClient(CLIENT_ID, manager).also { 
 // all the stuff
  client = it
}
```

Reply:
Well I guess for that, it could stay like it is